### ISSUE TITLE
[BUG] Error "Socket descriptor larger than FD_SETSIZE: 1163 > 1024"

### ISSUE THREAD

**Issue by kertor** [OpenSIPS/opensips#2457]
**OpenSIPS version you are running**

```
version: opensips 3.0.5 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-FUTEX-ADAPTIVE_WAIT, ENCRYPT-CFG
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 1b9c8cb12
main.c compiled on  with gcc 7
```

**Describe the bug**

I have a lot of `listen` processes and in this case command `opensips-cli -x mi get_statistics all` return error.
When i do few requests to opensips-cli i see error:
```
Apr  5 19:51:07 localhost opensips[5437]: Socket descriptor larger than FD_SETSIZE: 1163 > 1024
```

**To Reproduce**

1) just create many listeners with internal IP:port like
```
listen = udp:192.168.1.2:5061
listen = udp:192.168.1.2:5062
listen = udp:192.168.1.2:5063
listen = udp:192.168.1.2:5064
listen = udp:192.168.1.2:5065
...
listen = udp:192.168.1.2:5090
```
2) Run cli command:
```
opensips-cli -x mi get_statistics all
```
3) you will see error in opensips logs:
```
Apr  5 20:14:00 opensips opensips[3615549]: Socket descriptor larger than FD_SETSIZE: 1423 > 1024
Apr  5 20:14:00 opensips opensips[3615549]: Socket descriptor larger than FD_SETSIZE: 1423 > 1024
```

**Expected behavior**

opensips-cli return `get_statistics all` data without errors.

**Relevant System Logs**

Only these related erros on log level 4:
```
Apr  5 20:14:00 opensips opensips[3615549]: Socket descriptor larger than FD_SETSIZE: 1423 > 1024
Apr  5 20:14:00 opensips opensips[3615549]: Socket descriptor larger than FD_SETSIZE: 1423 > 1024
```

opensips-cli logs:
```
Traceback (most recent call last):
  File "/bin/opensips-cli", line 9, in <module>
    run_console()
  File "/bin/opensips-cli", line 6, in run_console
    main.main()
  File "/usr/lib/python3/dist-packages/opensipscli/main.py", line 77, in main
    shell = cli.OpenSIPSCLIShell(args)
  File "/usr/lib/python3/dist-packages/opensipscli/cli.py", line 85, in __init__
    self.update_instance(cfg.current_instance)
  File "/usr/lib/python3/dist-packages/opensipscli/cli.py", line 163, in update_instance
    self.modules[name] = (imod, mod.__get_methods__(imod))
  File "/usr/lib/python3/dist-packages/opensipscli/modules/mi.py", line 182, in __get_methods__
    return comm.execute('which')
  File "/usr/lib/python3/dist-packages/opensipscli/comm.py", line 40, in execute
    ret = comm_handler.execute(cmd, params)
  File "/usr/lib/python3/dist-packages/opensipscli/communication/http.py", line 38, in execute
    replycmd = urllib.request.urlopen(request).read().decode()
  File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/usr/lib/python3.8/urllib/request.py", line 542, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 1379, in http_open
    return self.do_open(http.client.HTTPConnection, req)
  File "/usr/lib/python3.8/urllib/request.py", line 1354, in do_open
    r = h.getresponse()
  File "/usr/lib/python3.8/http/client.py", line 1347, in getresponse
    response.begin()
  File "/usr/lib/python3.8/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.8/http/client.py", line 268, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 104] Connection reset by peer
```

**OS/environment information**
 - Operating System:  Ubuntu 20.04
 - OpenSIPS installation: 
 - other relevant information:

**Additional context**

Could you please check this?
Thank you very much for checking!

**Comment by bogdan-iancu** [comment]
Strange, that log does not look like one generated by OpenSIPS code. Maybe by a library. What MI backend do you use for the CLI ?

**Comment by kertor** [comment]
Hello Bogdan, thank you for response.

opensips-cli settings:
```
[main]
log_level: WARNING
prompt_name: main
prompt_intro: main
prompt_emptyline_repeat_cmd: False
history_file: ~/.opensips-cli-main.history
history_file_size: 1000
output_type: pretty-print
communication_type: http
url: http://192.168.1.2:8001/mi
```

Modules settings:
```
########################################################################
loadmodule "httpd.so"
########################################################################
modparam("httpd", "ip", "192.168.1.2")
modparam("httpd", "port", 8000)
modparam("httpd", "buf_size", 524288)

########################################################################
loadmodule "mi_http.so"
########################################################################
modparam("httpd", "ip", "192.168.1.2")
modparam("httpd", "port", 8001)
modparam("httpd", "buf_size", 524288) # Default 25% of pkg mem
```

**Comment by bogdan-iancu** [comment]
Somehow I have the feeling that the log comes from the microhttp library, 

By running an `opensips-cli -x mi ps`, could you confirm the name/type of the process with the same pid as in the culprit logs?

**Comment by kertor** [comment]
`opensips-cli -x mi ps` not working too, i see the same error.
But i see that problem PID from logs match to second opensip sprocess.
Second opensips process it usual `HTTPD`:
```
{
    "Processes": [
        {
            "ID": 0,
            "PID": 163275,
            "Type": "attendant"
        },
        {
            "ID": 1,
            "PID": 163276,
            "Type": "HTTPD 192.168.1.2:8001"
        },
```

**Comment by bogdan-iancu** [comment]
@kertor , some small progress here while investigating the libmicrohttpd (looking to migrate the module to the opensips reactor framework)
```
MHD_USE_POLL

    Use poll() instead of select(). This allows sockets with descriptors >= FD_SETSIZE. This option currently only works in conjunction with MHD_USE_INTERNAL_POLLING_THREAD (at this point). If you specify MHD_USE_POLL and the local platform does not support it, MHD_start_daemon will return NULL.
```
This is form https://www.gnu.org/software/libmicrohttpd/manual/html_node/microhttpd_002dconst.html
And currently this flag is not used, but it may be mandatory during the upcoming changes. 

I will keep you posted here.

**Comment by bogdan-iancu** [comment]
@kertor , could you try out this patch, it should solve your problem. Please confirm

[opensips_httpd_FDSETSIZE.patch.txt](https://github.com/OpenSIPS/opensips/files/6341288/opensips_httpd_FDSETSIZE.patch.txt)

**Comment by kertor** [comment]
Thank you Bogdan!
Will check and report about that.

**Comment by bogdan-iancu** [comment]
This is addressed in 3.2
